posts/inside-rust/2024-08-09-async-closures-call-for-testing.md

```diff
@@ -0,0 +1,124 @@
+---
+layout: post
+title: "Async Closures MVP: Call for Testing!"
+author: Michael Goulet
+team: The Async Working Group <https://www.rust-lang.org/governance/wgs/wg-async>
+---
+
+The async working group is excited to announce that [RFC 3668] "Async Closures" was recently approved by the Lang team. In this post, we want to briefly motivate why they exist, explain their current shortcomings, and most importantly, announce a call for testing them on nightly Rust.
+
+## The backstory
+
+Async closures were originally proposed in [RFC 2394](https://rust-lang.github.io/rfcs/2394-async_await.html#async--closures) which introduced `async`/`await` to the language. Simple handling of async closures has existed since async-await was implemented [soon thereafter](https://github.com/rust-lang/rust/pull/51580), but until recently async closures simply desugared into closures that returned async blocks:
+
+```rust
+let x = async || {};
+
+// ...is just sugar for:
+let x = || { async {} };
+```
+
+This had a fundamental limitation that it's impossible to express a closure that returns a future that borrows captured state.
+
+Somewhat relatedly, on the callee side, when users want to take an async closure as an argument, they must express that as a bound of two different generic types, or use boxing:
+
+```rust
+fn async_callback<F, Fut>(callback: F)
+where
+    F: FnOnce() -> Fut,
+    Fut: Future<Output = String>;
+
+// Or:
+fn async_callback_boxed<F, Fut>(callback: F)
+where
+    F: FnOnce() -> Pin<Box<dyn Future<Output = String>>>;
+```
+
+This led to an additional limitation, that it's impossible to express higher ranked async fn bounds without boxing, since a higher ranked trait bound on `F` cannot lead to a higher-ranked type for `Fut`.
```

Comment:
```suggestion
This led to an additional limitation, that it's impossible to express higher-ranked async fn bounds without boxing, since a higher-ranked trait bound on `F` cannot lead to a higher-ranked type for `Fut`.
```

(Or it could go the other way.)